utils/s3/client.cc

```diff
@@ -385,10 +403,110 @@ class client::upload_sink final : public client::upload_sink_base {
+    req.query_parameters["partNumber"] = format("{}", part_number);
+    req.query_parameters["uploadId"] = _upload_id;
+    req._headers["x-amz-copy-source"] = piece._object_name;
+
+    // See comment in upload_part(memory_data_sink_buffers) overload regarding the
+    // _flush_sem usage and _part_etags assignments
+    //
+    // Before the piece's object can be copied into the target one, it should be
+    // flushed and closed. After the object is copied, it can be removed. If copy
+    // goes wrong, the object should be removed anyway.
+    auto units = co_await get_units(_flush_sem, 1);
+    (void)piece.flush().then([&piece] () {
+        return piece.close();
+    }).then([this, part_number, req = std::move(req)] () mutable {
+        return _http.make_request(std::move(req), [this, part_number] (const http::reply& rep, input_stream<char>&& in_) mutable -> future<> {
+            auto in = std::move(in_);
+            auto body = co_await util::read_entire_stream_contiguous(in);
+            auto etag = parse_multipart_copy_upload_etag(body);
+            if (_upload_id.empty()) {
+                co_await coroutine::return_exception(std::runtime_error("cannot copy part upload"));
+            }
+            s3l.trace("copy-uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+            _part_etags[part_number] = std::move(etag);
+        }).handle_exception([this, part_number] (auto ex) {
+            // ... the exact exception only remains in logs
+            s3l.warn("couldn't copy-upload part {}: {} (upload id {})", part_number, ex, _upload_id);
+        });
+    }).finally([this, &piece] {
+        return _client->delete_object(piece._object_name).handle_exception([&piece] (auto ex) {
+            s3l.warn("failed to remove copy-upload piece {}", piece._object_name);
+        });
+    }).finally([units = std::move(units), piece_ptr = std::move(piece_ptr)] {});
+}
+
+class client::upload_jumbo_sink final : public upload_sink_base {
+    // "Part numbers can be any number from 1 to 10,000, inclusive."
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+    // Minus one here, because parts start counting from 1
+    static constexpr unsigned aws_maximum_parts_in_piece = 10000 - 1;
+
+    const unsigned _maximum_parts_in_piece;
+    std::unique_ptr<upload_sink> _current;
+
+    future<> maybe_flush() {
+        if (_current->parts_count() >= _maximum_parts_in_piece) {
+            auto next = std::make_unique<upload_sink>(_client, format("{}_{}", _object_name, parts_count() + 1));
+            co_await upload_part(std::exchange(_current, std::move(next)));
+            s3l.trace("Initiated {} piece (upload_id {})", parts_count(), _upload_id);
+        }
+    }
+
+public:
+    upload_jumbo_sink(shared_ptr<client> cln, sstring object_name, std::optional<unsigned> max_parts)
+        : upload_sink_base(std::move(cln), std::move(object_name))
```

Comment:
another source of confusion related to part number.

see above you defined parts_count() as the actual number of parts, which is _part_etags.size().

whereas "aws_maximum_parts_in_piece = 10000 - 1" refers to the maximum index = 9999.

if parts_count == 9999, you still want to allow another part inside the piece, to reach the maximum of 10k, no?

Reply:
@xemul I took a look at it again, and I think there actually might be an off-by-one.

When you construct upload_jumbo_sink() with optional max_parts param set to 3, you expect each object piece to contain 3 parts, right?

Now why the default which is supposed to be the max number of parts per object chunk defaults to 10k-1=9999?

What happens is that current upload_sink writes one part at a time for current object piece, and after each write, upload_jumbo_sink decides in maybe_flush() whether current piece should be uploaded as a part of the jumbo

With max set to 9999, upload_sink will be flushed as soon as current object piece has 9999 parts (not confuse it with part number). Therefore, an off-by-one.

To fix it, let's just set aws_maximum_parts_in_piece to 10k, not 10k-1. Makes sense?

Reply:
> With max set to 9999, upload_sink will be flushed as soon as current object piece has 9999 parts (not confuse it with part number). Therefore, an off-by-one.

If the piece contains 9999 parts, then their numbers are 1, 2, ... 10000, just as AWS documentation insists.

Reply:
If you do this change, 

```
@@ -451,13 +452,14 @@ class client::upload_jumbo_sink final : public upload_sink_base {
     // "Part numbers can be any number from 1 to 10,000, inclusive."
     // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
     // Minus one here, because parts start counting from 1
-    static constexpr unsigned aws_maximum_parts_in_piece = 10000 - 1;
+    static constexpr unsigned aws_maximum_parts_in_piece = 10 - 1;
```

and tweak test_client_multipart_copy_upload to not specify the max pieces (== std::nullopt), you'll probably see each object piece containing only 9 parts, not 10.

Reply:
> ... you'll probably see each object piece containing only 9 parts, not 10.

Yes, because 10000 is not the number of pieces I want, it's the last number of piece. While the change implies -- I want 10 pieces instead of 10000, that's not correct.

Reply:
But don't you want 10k parts per piece? That's the goal, right? To use the full capacity for each object piece. You don't want to limit each to 10k-1 parts.

`_current->parts_count() `refers to the number of parts count of current object piece, which starts from 0.

With `aws_maximum_parts_in_piece = 10000 - 1 = 9999
`
We should start a new object piece

when _current->parts_count() is 10k, in the check ` (_current->parts_count() >= _maximum_parts_in_piece)`

If _maximum_parts_in_piece is set to 9999, I start a new object piece when the current one has only 9999 parts.

As for the part_number sent to aws s3, that's not a problem. The upload_sink for the current object piece already wrote all the 10k parts using the correct part number, using the range [1, 10k]. 

client::upload_jumbo_sink::maybe_flush() is called after the parts were uploaded to the current object piece.

Reply:
> > ... you'll probably see each object piece containing only 9 parts, not 10.
> 
> Yes, because 10000 is not the number of pieces I want, it's the last number of piece. While the change implies -- I want 10 pieces instead of 10000, that's not correct.

Let's remember that you still want each piece to have 10k parts. Not 9999.

Reply:
with this change:

```
@@ -67,8 +68,9 @@ void do_test_client_multipart_upload(bool with_copy_upload) {
     auto out = output_stream<char>(
         // Make it 3 parts per piece, so that 128Mb buffer below
         // would be split into several 15Mb pieces
-        with_copy_upload ? cln->make_upload_jumbo_sink(name, 3) : cln->make_upload_sink(name)
+        with_copy_upload ? cln->make_upload_jumbo_sink(name, std::nullopt) : cln->make_upload_sink(name)
```

and this change:

```
@@ -451,13 +452,14 @@ class client::upload_jumbo_sink final : public upload_sink_base {
     // "Part numbers can be any number from 1 to 10,000, inclusive."
     // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
     // Minus one here, because parts start counting from 1
-    static constexpr unsigned aws_maximum_parts_in_piece = 10000 - 1;
+    static constexpr unsigned aws_maximum_parts_in_piece = 10 - 1;
```

we have:

INFO  2023-04-19 14:35:49,755 [shard  0] testlog - Upload object (with copy = true)

```
INFO  2023-04-19 14:35:49,757 [shard  0] s3 - 	 PUT part 0, total =1
INFO  2023-04-19 14:35:49,759 [shard  0] s3 - 	 PUT part 1, total =2
INFO  2023-04-19 14:35:49,760 [shard  0] s3 - 	 PUT part 2, total =3
INFO  2023-04-19 14:35:49,761 [shard  0] s3 - 	 PUT part 3, total =4
INFO  2023-04-19 14:35:49,789 [shard  0] s3 - 	 PUT part 4, total =5
INFO  2023-04-19 14:35:49,791 [shard  0] s3 - 	 PUT part 5, total =6
INFO  2023-04-19 14:35:49,792 [shard  0] s3 - 	 PUT part 6, total =7
INFO  2023-04-19 14:35:49,814 [shard  0] s3 - 	 PUT part 7, total =8
INFO  2023-04-19 14:35:49,816 [shard  0] s3 - 	 PUT part 8, total =9
INFO  2023-04-19 14:35:49,817 [shard  0] s3 - jumbo: parts=0, max_parts=9; current: parts=9
INFO  2023-04-19 14:35:49,821 [shard  0] s3 - 	 PUT part 0, total =1
INFO  2023-04-19 14:35:49,822 [shard  0] s3 - 	 PUT part 1, total =2
INFO  2023-04-19 14:35:49,823 [shard  0] s3 - 	 PUT part 2, total =3
INFO  2023-04-19 14:35:49,826 [shard  0] s3 - 	 PUT part 3, total =4
INFO  2023-04-19 14:35:49,858 [shard  0] s3 - 	 PUT part 4, total =5
INFO  2023-04-19 14:35:49,860 [shard  0] s3 - 	 PUT part 5, total =6
INFO  2023-04-19 14:35:49,862 [shard  0] s3 - 	 PUT part 6, total =7
INFO  2023-04-19 14:35:49,892 [shard  0] s3 - 	 PUT part 7, total =8
INFO  2023-04-19 14:35:49,895 [shard  0] s3 - 	 PUT part 8, total =9
INFO  2023-04-19 14:35:49,898 [shard  0] s3 - jumbo: parts=1, max_parts=9; current: parts=9
INFO  2023-04-19 14:35:49,901 [shard  0] s3 - 	 PUT part 0, total =1
INFO  2023-04-19 14:35:49,903 [shard  0] s3 - 	 PUT part 1, total =2
INFO  2023-04-19 14:35:49,905 [shard  0] s3 - 	 PUT part 2, total =3
INFO  2023-04-19 14:35:49,906 [shard  0] s3 - 	 PUT part 3, total =4
INFO  2023-04-19 14:35:49,947 [shard  0] s3 - 	 PUT part 4, total =5
INFO  2023-04-19 14:35:49,949 [shard  0] s3 - 	 PUT part 5, total =6

```

each piece have a maximum of 9 parts, even though we could have allowed them to grow to 10.

Reply:
> Let's remember that you still want each piece to have 10k parts. Not 9999.

No I don't. I want each piece to have part numbers stop at 10000. Given they start at 1, the maximum number of parts in a piece is 9999.

Reply:
I am thinking about the limit of 10k. I just reduce it to 10 for making easier to reproduce what it would have happened once input data is big enough to use the 50G for each object piece.

Reply:
> No I don't. I want each piece to have part numbers stop at 10000. Given they start at 1, the maximum number of parts in a piece is 9999.

Wait... Maybe I do

Reply:
> > Let's remember that you still want each piece to have 10k parts. Not 9999.
> 
> No I don't. I want each piece to have part numbers stop at 10000. Given they start at 1, the maximum number of parts in a piece is 9999.

range [1, 10k] with upper bound included, allows you to have 10k parts. the range size is 10k. what am I missing? :-)

Reply:
:man_facepalming: Yes, it's off-by-one. Parts are numbered 1, 2, ..., 10000 and there are 10000 of them.

Reply:
We should print this conversation as a strong evidence that off-by-one is one of the hardest problems in computer science haha